src/parallel/stderr.rs

```diff
@@ -0,0 +1,84 @@
+/// Helpers functions for [ChildStderr].
+use std::{io::Read, process::ChildStderr};
+
+use crate::Error;
+
+#[cfg(all(not(unix), not(windows)))]
+compile_error!("Only unix and windows support non-blocking pipes! For other OSes, disable the parallel feature.");
+
+#[cfg(unix)]
+fn get_flags(fd: &std::os::unix::io::RawFd) -> Result<libc::c_int, Error> {
+    let flags = unsafe { libc::fcntl(*fd, libc::F_GETFL, 0) };
+    if flags < 0 {
+        return Err(Error::new(
+            crate::ErrorKind::IOError,
+            format!(
+                "Failed to get flags for child stderr: {}",
+                std::io::Error::last_os_error()
+            ),
+        ));
+    }
+    Ok(flags)
+}
+
+#[allow(unused_variables)]
+pub fn set_non_blocking(stderr: &mut ChildStderr) -> Result<(), Error> {
+    // On Unix, switch the pipe to non-blocking mode.
+    // On Windows, we have a different way to be non-blocking.
+    #[cfg(unix)]
+    {
+        use std::os::unix::io::AsRawFd;
+        let fd = stderr.as_raw_fd();
+        let flags = get_flags(&fd)?;
+
+        if unsafe { libc::fcntl(fd, libc::F_SETFL, flags | libc::O_NONBLOCK) } != 0 {
+            return Err(Error::new(
+                crate::ErrorKind::IOError,
+                format!(
+                    "Failed to set flags for child stderr: {}",
+                    std::io::Error::last_os_error()
+                ),
+            ));
+        }
+    }
+
+    Ok(())
+}
+
+pub fn read_non_blocked(stderr: &mut ChildStderr, buf: &mut [u8]) -> Result<usize, std::io::Error> {
+    // On Windows, take a peek at the pipe to see if there's data.
+    #[cfg(windows)]
+    {
+        use crate::windows::windows_sys::PeekNamedPipe;
+        use std::os::windows::io::AsRawHandle;
+        use std::ptr::null_mut;
+        let mut bytes_available = 0;
+        unsafe {
+            if PeekNamedPipe(
+                stderr.as_raw_handle(),
+                null_mut(),
+                0,
+                null_mut(),
+                &mut bytes_available,
+                null_mut(),
+            ) == 0
+            {
+                return Err(std::io::Error::last_os_error());
+            }
+        }
+        if bytes_available == 0 {
+            return Err(std::io::Error::new(
```

Comment:
Recommend  to use <io::Error as From> impl for io::ErrorKind, io::Error::new allocates internally while the from errorkind does not.